launch/config.py

```
@@ -0,0 +1,385 @@
+                if kk.endswith('_filename'):
+                    config[k][kk] = expand_path(vv, config_dir)
+        if isinstance(config[k], dict):
+            config[k] = yaml.dump(config[k])
+    config = stringify_configuration(config)
+    return validate_config(config)
+
+
+def get_user_argument_source(user_config: dict):
+    """ Unlike gen, nested config are allows here
+    """
+    user_source = Source(is_user=True)
+    for name, value in user_config.items():
+        user_source.add_must(name, value)
+    return user_source
+
+
+def validate_config(user_config: dict):
+    user_source = get_user_argument_source(user_config)
+    t = get_target()
+    sources = [get_source(), user_source]
+    resolver = resolve_configuration(sources, [t])
+    status = resolver.status_dict
+
+    if status['status'] == 'errors':
+        raise launch.util.LauncherError('ValidationError', pretty_print_validate_error(status))
+
+    late_variables = dict()
+
+    for source in sources:
+        for setter_list in source.setters.values():
+            for setter in setter_list:
+                if not setter.is_late:
+                    continue
+
+                if setter.name not in resolver.late:
+                    continue
+
+                # Skip late vars that aren't referenced by config.
+                if not resolver.arguments[setter.name].is_finalized:
+                    continue
+
+                # Validate a late variable should only have one source.
+                assert setter.name not in late_variables
+
+                late_variables[setter.name] = setter.late_expression
+
+    final_args = {k: v.value for k, v in resolver.arguments.items() if v.is_finalized}
+    unrecognized_args = set(user_config.keys()) - set(final_args.keys())
+    if len(unrecognized_args) > 0:
+        raise launch.util.LauncherError(
+            'ValidationError', 'Unrecognized/incompatible arguments: {}'.format(unrecognized_args))
```

That gen silently accepts unrecognized parameters is IMO a bug, and should be fixed there. It's caused some user confusion in the past, though I can't find a ticket for it. IIRC gen used to error when it saw an unrecognized parameter, but at some point that was broken and never got fixed.